File: Sources/Services/AIService.swift
import Foundation

// MARK: - AI Service
class AIService: ObservableObject {
    static let shared = AIService()
    
    private init() {}
    
    // MARK: - Public Methods
    func getBestMove(for board: [[Player]], difficulty: Difficulty, player: Player) -> Position? {
        switch difficulty {
        case .easy:
            return getRandomMove(board: board)
        case .medium:
            return Bool.random() ? getRandomMove(board: board) : getMinimaxMove(board: board, player: player, depth: 3)
        case .hard:
            return getMinimaxMove(board: board, player: player, depth: 6)
        }
    }
    
    // MARK: - Private Methods
    
    /// Easy AI: Returns a random valid move
    private func getRandomMove(board: [[Player]]) -> Position? {
        var availablePositions: [Position] = []
        
        for row in 0..<3 {
            for column in 0..<3 {
                if board[row][column] == .none {
                    availablePositions.append(Position(row: row, column: column))
                }
            }
        }
        
        return availablePositions.randomElement()
    }
    
    /// Medium/Hard AI: Uses Minimax algorithm
    private func getMinimaxMove(board: [[Player]], player: Player, depth: Int) -> Position? {
        var bestScore = Int.min
        var bestMove: Position?
        
        let availablePositions = getAvailablePositions(board: board)
        
        for position in availablePositions {
            var newBoard = board
            newBoard[position.row][position.column] = player
            
            let score = minimax(
                board: newBoard,
                depth: depth - 1,
                isMaximizing: false,
                player: player,
                alpha: Int.min,
                beta: Int.max
            )
            
            if score > bestScore {
                bestScore = score
                bestMove = position
            }
        }
        
        return bestMove
    }
    
    /// Minimax algorithm with alpha-beta pruning
    private func minimax(
        board: [[Player]],
        depth: Int,
        isMaximizing: Bool,
        player: Player,
        alpha: Int,
        beta: Int
    ) -> Int {
        // Check for terminal states
        if let winner = checkWinner(board: board) {
            return winner == player ? 10 + depth : -10 - depth
        }
        
        if depth == 0 || isBoardFull(board: board) {
            return 0
        }
        
        let currentPlayer = isMaximizing ? player : player.opposite
        
        if isMaximizing {
            var maxEval = Int.min
            let availablePositions = getAvailablePositions(board: board)
            
            for position in availablePositions {
                var newBoard = board
                newBoard[position.row][position.column] = currentPlayer
                
                let eval = minimax(
                    board: newBoard,
                    depth: depth - 1,
                    isMaximizing: false,
                    player: player,
                    alpha: alpha,
                    beta: beta
                )
                
                maxEval = max(maxEval, eval)
                alpha = max(alpha, eval)
                
                if beta <= alpha {
                    break
                }
            }
            
            return maxEval
        } else {
            var minEval = Int.max
            let availablePositions = getAvailablePositions(board: board)
            
            for position in availablePositions {
                var newBoard = board
                newBoard[position.row][position.column] = currentPlayer
                
                let eval = minimax(
                    board: newBoard,
                    depth: depth - 1,
                    isMaximizing: true,
                    player: player,
                    alpha: alpha,
                    beta: beta
                )
                
                minEval = min(minEval, eval)
                beta = min(beta, eval)
                
                if beta <= alpha {
                    break
                }
            }
            
            return minEval
        }
    }
    
    /// Get all available positions on the board
    private func getAvailablePositions(board: [[Player]]) -> [Position] {
        var positions: [Position] = []
        
        for row in 0..<3 {
            for column in 0..<3 {
                if board[row][column] == .none {
                    positions.append(Position(row: row, column: column))
                }
            }
        }
        
        return positions
    }
    
    /// Check if there's a winner
    private func checkWinner(board: [[Player]]) -> Player? {
        // Check rows
        for row in 0..<3 {
            if board[row][0] != .none &&
               board[row][0] == board[row][1] &&
               board[row][1] == board[row][2] {
                return board[row][0]
            }
        }
        
        // Check columns
        for column in 0..<3 {
            if board[0][column] != .none &&
               board[0][column] == board[1][column] &&
               board[1][column] == board[2][column] {
                return board[0][column]
            }
        }
        
        // Check diagonals
        if board[1][1] != .none {
            // Top-left to bottom-right diagonal
            if board[0][0] == board[1][1] &&
               board[1][1] == board[2][2] {
                return board[1][1]
            }
            
            // Top-right to bottom-left diagonal
            if board[0][2] == board[1][1] &&
               board[1][1] == board[2][0] {
                return board[1][1]
            }
        }
        
        return nil
    }
    
    /// Check if the board is full
    private func isBoardFull(board: [[Player]]) -> Bool {
        for row in 0..<3 {
            for column in 0..<3 {
                if board[row][column] == .none {
                    return false
                }
            }
        }
        return true
    }
}

// MARK: - AI Utility Extensions
extension AIService {
    
    /// Evaluate board position for strategic analysis
    func evaluateBoard(board: [[Player]], player: Player) -> Int {
        var score = 0
        
        // Center position preference
        if board[1][1] == player {
            score += 3
        }
        
        // Corner positions preference
        let corners = [(0,0), (0,2), (2,0), (2,2)]
        for (row, col) in corners {
            if board[row][col] == player {
                score += 2
            }
        }
        
        // Edge positions (less valuable)
        let edges = [(0,1), (1,0), (1,2), (2,1)]
        for (row, col) in edges {
            if board[row][col] == player {
                score += 1
            }
        }
        
        return score
    }
    
    /// Check for potential winning moves
    func getWinningMove(board: [[Player]], player: Player) -> Position? {
        let availablePositions = getAvailablePositions(board: board)
        
        for position in availablePositions {
            var newBoard = board
            newBoard[position.row][position.column] = player
            
            if checkWinner(board: newBoard) == player {
                return position
            }
        }
        
        return nil
    }
    
    /// Check for potential blocking moves
    func getBlockingMove(board: [[Player]], player: Player) -> Position? {
        return getWinningMove(board: board, player: player.opposite)
    }
}

// MARK: - AI Strategy Heuristics
extension AIService {
    
    /// Advanced evaluation for medium difficulty
    func getMediumDifficultyMove(board: [[Player]], player: Player) -> Position? {
        // Priority 1: Win if possible
        if let winningMove = getWinningMove(board: board, player: player) {
            return winningMove
        }
        
        // Priority 2: Block opponent from winning
        if let blockingMove = getBlockingMove(board: board, player: player) {
            return blockingMove
        }
        
        // Priority 3: Take center if available
        if board[1][1] == .none {
            return Position(row: 1, column: 1)
        }
        
        // Priority 4: Take corners
        let corners = [(0,0), (0,2), (2,0), (2,2)]
        let availableCorners = corners.filter { board[$0.0][$0.1] == .none }
        if let corner = availableCorners.randomElement() {
            return Position(row: corner.0, column: corner.1)
        }
        
        // Priority 5: Take any available position
        return getRandomMove(board: board)
    }
}

File: lib/services/ai_service.dart
import 'dart:math' as math;
import 'game_models.dart';

class AIService {
  static final AIService _instance = AIService._internal();
  factory AIService() => _instance;
  AIService._internal();

  Position? getBestMove(
    List<List<Player>> board,
    Player player,
    Difficulty difficulty,
  ) {
    switch (difficulty) {
      case Difficulty.easy:
        return _getRandomMove(board);
      case Difficulty.medium:
        return math.Random().nextBool() 
            ? _getRandomMove(board) 
            : _getMinimaxMove(board, player, difficulty.depth);
      case Difficulty.hard:
        return _getMinimaxMove(board, player, difficulty.depth);
    }
  }

  Position? _getRandomMove(List<List<Player>> board) {
    final availablePositions = <Position>[];
    
    for (int row = 0; row < 3; row++) {
      for (int col = 0; col < 3; col++) {
        if (board[row][col] == Player.none) {
          availablePositions.add(Position(row, col));
        }
      }
    }
    
    if (availablePositions.isEmpty) return null;
    
    return availablePositions[math.Random().nextInt(availablePositions.length)];
  }

  Position? _getMinimaxMove(List<List<Player>> board, Player player, int depth) {
    Position? bestMove;
    int bestScore = -math.pow(10, 9).toInt();
    
    final availablePositions = <Position>[];
    for (int row = 0; row < 3; row++) {
      for (int col = 0; col < 3; col++) {
        if (board[row][col] == Player.none) {
          availablePositions.add(Position(row, col));
        }
      }
    }
    
    for (final position in availablePositions) {
      final newBoard = _copyBoard(board);
      newBoard[position.row][position.column] = player;
      
      final score = _minimax(
        newBoard,
        depth - 1,
        false,
        player,
        player.opposite,
        -math.pow(10, 9).toInt(),
        math.pow(10, 9).toInt(),
      );
      
      if (score > bestScore) {
        bestScore = score;
        bestMove = position;
      }
    }
    
    return bestMove;
  }

  int _minimax(
    List<List<Player>> board,
    int depth,
    bool isMaximizing,
    Player originalPlayer,
    Player currentPlayer,
    int alpha,
    int beta,
  ) {
    // Check terminal states
    final winner = _checkWinner(board);
    if (winner != Player.none) {
      return winner == originalPlayer ? 10 + depth : -10 - depth;
    }
    
    if (depth == 0 || _isBoardFull(board)) {
      return 0;
    }
    
    if (isMaximizing) {
      int maxEval = -math.pow(10, 9).toInt();
      
      for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 3; col++) {
          if (board[row][col] == Player.none) {
            final newBoard = _copyBoard(board);
            newBoard[row][col] = currentPlayer;
            
            final eval = _minimax(
              newBoard,
              depth - 1,
              false,
              originalPlayer,
              currentPlayer.opposite,
              alpha,
              beta,
            );
            
            maxEval = math.max(maxEval, eval);
            alpha = math.max(alpha, eval);
            
            if (beta <= alpha) {
              break; // Alpha-beta pruning
            }
          }
        }
      }
      
      return maxEval;
    } else {
      int minEval = math.pow(10, 9).toInt();
      
      for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 3; col++) {
          if (board[row][col] == Player.none) {
            final newBoard = _copyBoard(board);
            newBoard[row][col] = currentPlayer;
            
            final eval = _minimax(
              newBoard,
              depth - 1,
              true,
              originalPlayer,
              currentPlayer.opposite,
              alpha,
              beta,
            );
            
            minEval = math.min(minEval, eval);
            beta = math.min(beta, eval);
            
            if (beta <= alpha) {
              break; // Alpha-beta pruning
            }
          }
        }
      }
      
      return minEval;
    }
  }

  Player _checkWinner(List<List<Player>> board) {
    // Check rows
    for (int row = 0; row < 3; row++) {
      if (board[row][0] != Player.none &&
          board[row][0] == board[row][1] &&
          board[row][1] == board[row][2]) {
        return board[row][0];
      }
    }
    
    // Check columns
    for (int col = 0; col < 3; col++) {
      if (board[0][col] != Player.none &&
          board[0][col] == board[1][col] &&
          board[1][col] == board[2][col]) {
        return board[0][col];
      }
    }
    
    // Check diagonals
    if (board[1][1] != Player.none) {
      // Main diagonal
      if (board[0][0] == board[1][1] && board[1][1] == board[2][2]) {
        return board[1][1];
      }
      
      // Anti-diagonal
      if (board[0][2] == board[1][1] && board[1][1] == board[2][0]) {
        return board[1][1];
      }
    }
    
    return Player.none;
  }

  bool _isBoardFull(List<List<Player>> board) {
    for (int row = 0; row < 3; row++) {
      for (int col = 0; col < 3; col++) {
        if (board[row][col] == Player.none) {
          return false;
        }
      }
    }
    return true;
  }

  List<List<Player>> _copyBoard(List<List<Player>> board) {
    return board.map((row) => row.toList()).toList();
  }

  // Medium difficulty strategy improvements
  Position? getMediumDifficultyMove(
    List<List<Player>> board,
    Player player,
  ) {
    // Priority 1: Win if possible
    final winningMove = _findWinningMove(board, player);
    if (winningMove != null) return winningMove;
    
    // Priority 2: Block opponent from winning
    final blockingMove = _findWinningMove(board, player.opposite);
    if (blockingMove != null) return blockingMove;
    
    // Priority 3: Take center
    if (board[1][1] == Player.none) {
      return const Position(1, 1);
    }
    
    // Priority 4: Take corners
    final corners = [
      const Position(0, 0),
      const Position(0, 2),
      const Position(2, 0),
      const Position(2, 2),
    ];
    
    final availableCorners = corners.where((pos) => 
      board[pos.row][pos.column] == Player.none
    ).toList();
    
    if (availableCorners.isNotEmpty) {
      return availableCorners[math.Random().nextInt(availableCorners.length)];
    }
    
    // Priority 5: Take any available position
    return _getRandomMove(board);
  }

  Position? _findWinningMove(List<List<Player>> board, Player player) {
    for (int row = 0; row < 3; row++) {
      for (int col = 0; col < 3; col++) {
        if (board[row][col] == Player.none) {
          final testBoard = _copyBoard(board);
          testBoard[row][col] = player;
          
          if (_checkWinner(testBoard) == player) {
            return Position(row, col);
          }
        }
      }
    }
    return null;
  }
}